openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,102 @@
+Discussion provider configuration
+=================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+As part of the BD-03 initiative (Blended Development, Project 3),
+we want to enable third-party discussion providers to replace the
+default forums experience in edx-platform.
+
+To accomplish this, we should establish a configuration system to
+enable/disable/configure these new discussion-provider plugins;
+the existing/default/legacy forums experience will be included as one of these plugins.
+
+This ADR proposes a new configuration system that allows operators to create
+pre-populated configurations for specific discussions providers,
+which can then be used and customized by course authors/admins.
+
+
+Decision
+--------
+
+We propose to implement this as a Django database-backed model with
+historical records, `LearningContextDiscussionConfig`.
+
+While the underlying model will use `LearningContextKeys` (so that this
+could eventually be extended to cover (sites/programs/etc.), the first
+implementing data structure (`data.py`) can be described as:
+
+    @attr.s(frozen=True)
+    class CourseDiscussionConfigData:
+        course_key = attr.ib(type=CourseKey)
+        enabled = attr.ib(type=bool)
+        lti_configuration = attr.ib(type=dict)
+        plugin_configuration = attr.ib(type=dict)
+        provider = attr.ib(type=str)
+
+Discussions can be disabled on a course by setting `enabled=False`;
+if no record exists, the plugin is considered to be disabled by default.
+
+`provider` is the (arbitrary, but stable) key/id for the discussion provider.
+
+`plugin_configuration` is a free-form dictionary of configuration values,
+to be passed to plugin's renderer
+
+`lti_configuration` is a collection of LTI launch configuration data
+(URL, key, secret, etc.), but this data structure should be extracted
+into its own model; see notes below (`Dependent Work`).
+
+Changes to this model are tracked using Django Simple History.
+
+Corressponding getters/setters/helpers will be created to provide access.
+
+
+Dependent Work
+--------------
+
+The configuration model could start by storing the LTI launch
+configuration as a simple dictionary on the model/record, but ultimately
+(initially?), this configuration should be extracted into its own
+table/model (strictly for LTI launch config) and then linked to from
```

Comment:
How much additional effort is this abstraction work? I see the benefit but would like to know the cost

Reply:
Had conversation with @stvstnfrd about this, and (pending @xitij2000 thoughts) we'd like to go ahead and merge this as a simple dictionary and migrate that into its own table/model later on, perhaps quite soon after this merges. That will help us move forward more quickly.

Reply:
Change made. @xitij2000 Can you clarify what, if any, feature gaps exist w/ the `LtiConfiguration` model?

Reply:
Currently that model can only store LTI 1.3 credentials. I have a task this sprint to update it to support 1.1 as well based on the approach outlined by @giovannicimolin